File: xNet/src/main/java/org/sqlunet/style/Spanner.java
/*
 * Copyright (c) 2023. Bernard Bou
 */

package org.sqlunet.style;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Paint.FontMetricsInt;
import android.graphics.drawable.Drawable;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.style.ClickableSpan;
import android.text.style.DynamicDrawableSpan;
import android.text.style.ImageSpan;
import android.text.style.ReplacementSpan;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import org.sqlunet.xnet.R;

import java.util.Collection;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.content.res.AppCompatResources;

/**
 * Spanner
 *
 * @author <a href="mailto:1313ou@gmail.com">Bernard Bou</a>
 */
public class Spanner
{
	static private final String TAG = "Spanner";

	/**
	 * Collapsed marker
	 */
	static private final char COLLAPSEDCHAR = '@';

	/**
	 * Collapsed marker
	 */
	static private final String COLLAPSEDSTRING = Character.toString(COLLAPSEDCHAR);

	/**
	 * Expanded marker
	 */
	static private final String EXPANDEDSTRING = "~";
	/**
	 * End of expanded string marker
	 */
	static private final char EOEXPANDEDSTRING = '~';

	// I N T E R F A C E S

	/**
	 * Span factory
	 *
	 * @author <a href="mailto:1313ou@gmail.com">Bernard Bou</a>
	 */
	@FunctionalInterface
	public interface SpanFactory
	{
		@Nullable
		Object make(@SuppressWarnings("UnusedParameters") final long flags);
	}

	/**
	 * Click image interface
	 *
	 * @author <a href="mailto:1313ou@gmail.com">Bernard Bou</a>
	 */
	@FunctionalInterface
	public interface OnClickImage
	{
		void onClickImage(final SpannableStringBuilder sb, final int position, final boolean collapsed);
	}

	// H I D D E N S P A N

	/**
	 * Hidden span
	 *
	 * @author <a href="mailto:1313ou@gmail.com">Bernard Bou</a>
	 */
	static public class HiddenSpan extends ReplacementSpan
	{
		/**
		 * Constructor
		 */
		public HiddenSpan()
		{
		}

		@Override
		public void draw(@NonNull final Canvas canvas, final CharSequence arg1, final int arg2, final int arg3, final float arg4, final int arg5, final int arg6, final int arg7, @NonNull final Paint arg8)
		{
			//
		}

		@Override
		public int getSize(@NonNull final Paint paint, final CharSequence text, final int from, final int to, final FontMetricsInt fm)
		{
			return 0;
		}
	}

	// A P P L Y S P A N

	/**
	 * Apply spans
	 *
	 * @param sb    spannable string builder
	 * @param from  start
	 * @param to    finish
	 * @param spans spans to apply
	 */
	@SuppressWarnings("unchecked")
	static void setSpan(@NonNull @SuppressWarnings("TypeMayBeWeakened") final SpannableStringBuilder sb, final int from, final int to, @Nullable final Object spans)
	{
		if (spans != null && to - from > 0)
		{
			if (spans instanceof Object[])
			{
				for (final Object span : (Object[]) spans)
				{
					if (span != null)
					{
						sb.setSpan(span, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
					}
				}
			}
			else if (spans instanceof Collection)
			{
				for (Object span2 : (Collection<Object>) spans)
				{
					sb.setSpan(span2, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
				}
			}
			else
			{
				sb.setSpan(spans, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
			}
		}
	}

	/**
	 * Apply spans
	 *
	 * @param sb        spannable string builder
	 * @param from      start
	 * @param to        finish
	 * @param flags     flags
	 * @param factories span factories to call to get spans
	 */
	static public void setSpan(@NonNull final SpannableStringBuilder sb, final int from, final int to, @SuppressWarnings("SameParameterValue") final long flags, @NonNull final SpanFactory... factories)
	{
		if (to - from > 0)
		{
			for (final SpanFactory spanFactory : factories)
			{
				final Object spans = spanFactory.make(flags);
				Spanner.setSpan(sb, from, to, spans);
			}
		}
	}

	// I M A G E

	/**
	 * Append spans
	 *
	 * @param sb    spannable string builder
	 * @param spans image span with possible image style span
	 */
	static private void appendImageSpans(@NonNull final SpannableStringBuilder sb, @NonNull final Object... spans)
	{
		final int from = sb.length();
		sb.append(COLLAPSEDCHAR);
		final int to = sb.length();
		for (Object span : spans)
		{
			sb.setSpan(span, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
		}
	}

	/**
	 * Append image
	 *
	 * @param sb       spannable string builder
	 * @param drawable drawable to use
	 */
	static public void appendImage(@NonNull final SpannableStringBuilder sb, @NonNull final Drawable drawable)
	{
		final Object span = new ImageSpan(drawable, DynamicDrawableSpan.ALIGN_BASELINE);
		Spanner.appendImageSpans(sb, span);
	}

	// C L I C K A B L E

	/**
	 * Append clickable image
	 *
	 * @param sb       spannable string builder
	 * @param caption  caption
	 * @param listener click listener
	 * @param context  context
	 */
	static public void appendClickableImage(@NonNull final SpannableStringBuilder sb, @NonNull final CharSequence caption, @NonNull final OnClickImage listener, @NonNull final Context context)
	{
		final Drawable collapsedDrawable = getDrawable(context, R.drawable.ic_collapsed);
		final Drawable expandedDrawable = getDrawable(context, R.drawable.ic_expanded);
		appendClickableImage(sb, collapsedDrawable, expandedDrawable, caption, listener);
	}

	/**
	 * Append clickable image
	 *
	 * @param sb                spannable string builder
	 * @param collapsedDrawable collapse drawable
	 * @param expandedDrawable  expandContainer drawable
	 * @param caption           caption
	 * @param listener          click listener
	 */
	static private void appendClickableImage(@NonNull final SpannableStringBuilder sb, @NonNull final Drawable collapsedDrawable, final Drawable expandedDrawable, @NonNull final CharSequence caption, @NonNull final OnClickImage listener)
	{
		final ImageSpan span = new ImageSpan(collapsedDrawable, DynamicDrawableSpan.ALIGN_BASELINE);
		final ClickableSpan span2 = new ClickableSpan()
		{
			@Override
			synchronized public void onClick(@NonNull final View view)
			{
				// Log.d(TAG, "Click image");
				final TextView textView = (TextView) view;
				final SpannableStringBuilder sb1 = (SpannableStringBuilder) textView.getText();
				final int clickableStart = textView.getSelectionStart();
				final int clickableEnd = textView.getSelectionEnd();
				final Object[] spans = sb1.getSpans(clickableStart, clickableEnd, ImageSpan.class);
				for (final Object span3 : spans)
				{
					// get image span
					int from = sb1.getSpanStart(span3);
					int to = sb1.getSpanEnd(span3);

					// remove image span
					sb1.removeSpan(span3);

					// text
					final char c = sb1.charAt(from);
					final boolean collapsed = c == COLLAPSEDCHAR;
					sb1.replace(from, to, collapsed ? EXPANDEDSTRING : COLLAPSEDSTRING);

					// set new image span
					final Object newImageSpan = new ImageSpan(collapsed ? expandedDrawable : collapsedDrawable, DynamicDrawableSpan.ALIGN_BASELINE);
					sb1.setSpan(newImageSpan, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);

					// fire click
					Log.d(TAG, from + "->" + to);
					listener.onClickImage(sb1, to + caption.length() + 2, collapsed);
				}
				textView.setText(sb1);
			}
		};
		Spanner.appendImageSpans(sb, span, span2);
		sb.append(' ').append(caption).append('\n');
	}

	/**
	 * Insert tag
	 *
	 * @param sb       spannable string builder
	 * @param position insert position
	 * @param tag      tag
	 */
	static public void insertTag(@NonNull final SpannableStringBuilder sb, final int position, @NonNull final CharSequence tag)
	{
		final String insert = tag.toString() + '\n' + Spanner.EOEXPANDEDSTRING;
		sb.insert(position, insert);
		int mark = position + tag.length() + 1;
		sb.setSpan(new Spanner.HiddenSpan(), mark, mark + 1, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
	}

	/**
	 * Collapse
	 *
	 * @param sb       spannable string builder
	 * @param position position to start from (to end-of-expanded string)
	 */
	static public void collapse(@NonNull final SpannableStringBuilder sb, final int position)
	{
		sb.delete(position, Spanner.find(sb, position, Spanner.EOEXPANDEDSTRING));
	}

	// T E X T

	/**
	 * Append text
	 *
	 * @param sb        spannable string builder
	 * @param text      text
	 * @param flags     flags
	 * @param factories span factories
	 * @return input spannable string builder
	 */
	@NonNull
	static public Appendable append(@NonNull final SpannableStringBuilder sb, @Nullable final CharSequence text, @SuppressWarnings("SameParameterValue") final long flags, @Nullable final SpanFactory... factories)
	{
		if (text != null && text.length() > 0)
		{
			final int from = sb.length();
			sb.append(text);
			final int to = sb.length();
			if (factories != null)
			{
				for (final SpanFactory spanFactory : factories)
				{
					final Object span = spanFactory.make(flags);
					if (span != null)
					{
						applySpan(sb, from, to, span);
					}
				}
			}
		}
		return sb;
	}

	/**
	 * Append text
	 *
	 * @param sb    spannable string builder
	 * @param text  text
	 * @param spans spans to apply
	 */
	@NonNull
	static public SpannableStringBuilder appendWithSpans(@NonNull final SpannableStringBuilder sb, @Nullable final CharSequence text, @NonNull final Object... spans)
	{
		if (text != null && text.length() > 0)
		{
			final int from = sb.length();
			sb.append(text);
			final int to = sb.length();
			applySpans(sb, from, to, spans);
		}
		return sb;
	}

	/**
	 * Apply spans
	 *
	 * @param sb    spannable string builder
	 * @param from  from position
	 * @param to    to position
	 * @param spans spans to apply
	 */
	static public void applySpans(@NonNull final SpannableStringBuilder sb, final int from, final int to, @NonNull final Object... spans)
	{
		for (final Object span : spans)
		{
			applySpan(sb, from, to, span);
		}
	}

	/**
	 * Apply span
	 *
	 * @param sb   spannable string builder
	 * @param from from position
	 * @param to   to position
	 * @param span span to apply
	 */
	@SuppressWarnings("unchecked")
	static public void applySpan(@NonNull final SpannableStringBuilder sb, final int from, final int to, @NonNull final Object span)
	{
		if (span instanceof Object[])
		{
			for (Object span2 : (Object[]) span)
			{
				sb.setSpan(span2, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
			}
		}
		else if (span instanceof Collection)
		{
			for (Object span2 : (Collection<Object>) span)
			{
				sb.setSpan(span2, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
			}
		}
		else
		{
			sb.setSpan(span, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
		}
	}

	// H E L P E R S

	/**
	 * Find delimiter
	 *
	 * @param sb        spannable string builder
	 * @param start     search start
	 * @param delimiter delimiter
	 * @return delimiter position or -1 if not found
	 */
	static private int find(@NonNull @SuppressWarnings("TypeMayBeWeakened") final SpannableStringBuilder sb, final int start, @SuppressWarnings("SameParameterValue") final char delimiter)
	{
		int i = start;
		while (i < sb.length())
		{
			if (sb.charAt(i) == delimiter)
			{
				return i + 1;
			}
			i++;
		}
		return -1;
	}

	/**
	 * Get drawable from resource id
	 *
	 * @param context context
	 * @param resId   resource id
	 * @return drawable
	 */
	@NonNull
	static public Drawable getDrawable(@NonNull final Context context, @DrawableRes final int resId)
	{
		final Drawable drawable = AppCompatResources.getDrawable(context, resId);
		assert drawable != null;
		drawable.setBounds(0, 0, drawable.getIntrinsicWidth(), drawable.getIntrinsicHeight());
		return drawable;
	}

	// --Commented out by Inspection START (8/20/17 10:09 AM):
	//	/**
	//	 * Dump spannable string builder
	//	 *
	//	 * @param sb spannable string builder
	//	 * @return string
	//	 */
	//	static public String dump(final SpannableStringBuilder sb)
	//	{
	//		final StringBuilder dump = new StringBuilder();
	//		for (Object span : sb.getSpans(0, sb.length(), Object.class))
	//		{
	//			final int from = sb.getSpanStart(span);
	//			final int to = sb.getSpanEnd(span);
	//			final CharSequence sub = sb.subSequence(from, to);
	//			final Class<?> clazz = span.getClass();
	//			dump.append(sub).append(' ').append('[').append(from).append(',').append(to).append(']').append(' ').append(clazz).append(' ');
	//			if (clazz.equals(BackgroundColorSpan.class))
	//			{
	//				final BackgroundColorSpan bSpan = (BackgroundColorSpan) span;
	//				int color = bSpan.getBackgroundColor();
	//				dump.append(Integer.toHexString(color));
	//			}
	//			else if (clazz.equals(ForegroundColorSpan.class))
	//			{
	//				final ForegroundColorSpan fSpan = (ForegroundColorSpan) span;
	//				int color = fSpan.getForegroundColor();
	//				dump.append(Integer.toHexString(color));
	//			}
	//			else if (clazz.equals(ImageSpan.class))
	//			{
	//				final ImageSpan iSpan = (ImageSpan) span;
	//				dump.append(iSpan.getDrawable()).append(' ').append(iSpan.getSource());
	//			}
	//			else if (clazz.equals(HiddenSpan.class))
	//			{
	//				final HiddenSpan hSpan = (HiddenSpan) span;
	//				dump.append(hSpan.toString());
	//			}
	//			dump.append('\n');
	//		}
	//		return dump.toString();
	//	}
	// --Commented out by Inspection STOP (8/20/17 10:09 AM)
}

File: xNet/src/main/java/org/sqlunet/style/Spanner.kt
/*
 * Copyright (c) 2023. Bernard Bou
 */
package org.sqlunet.style

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Paint.FontMetricsInt
import android.graphics.drawable.Drawable
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.TextPaint
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.text.style.DynamicDrawableSpan
import android.text.style.ImageSpan
import android.text.style.ReplacementSpan
import android.util.Log
import android.view.View
import android.widget.TextView
import androidx.annotation.DrawableRes
import androidx.annotation.ReturnThis
import androidx.appcompat.content.res.AppCompatResources
import org.sqlunet.xnet.R

typealias Span = Any

/**
 * Spanner
 *
 * @author [Bernard Bou](mailto:1313ou@gmail.com)
 */
open class Spanner {

    // I N T E R F A C E S

    /**
     * Span factory
     *
     * @author [Bernard Bou](mailto:1313ou@gmail.com)
     */
    fun interface SpanFactory {

        fun make(flags: Long): Span?
    }

    // H I D D E N S P A N

    /**
     * Hidden span
     *
     * @author [Bernard Bou](mailto:1313ou@gmail.com)
     */
    class HiddenSpan : ReplacementSpan() {

        override fun draw(canvas: Canvas, arg1: CharSequence, arg2: Int, arg3: Int, arg4: Float, arg5: Int, arg6: Int, arg7: Int, arg8: Paint) {
        }

        override fun getSize(paint: Paint, text: CharSequence, from: Int, to: Int, fm: FontMetricsInt?): Int {
            return 0
        }
    }

    companion object {

        private const val TAG = "Spanner"

        /**
         * Collapsed marker
         */
        private const val COLLAPSEDCHAR = '@'

        /**
         * Collapsed marker
         */
        private const val COLLAPSEDSTRING = COLLAPSEDCHAR.toString()

        /**
         * Expanded marker
         */
        private const val EXPANDEDSTRING = "~"

        /**
         * End of expanded string marker
         */
        private const val EOEXPANDEDSTRING = '~'

        // A P P L Y  S P A N

        /**
         * Apply spans
         *
         * @param sb    spannable string builder
         * @param from  start
         * @param to    finish
         * @param spans spans to apply
         */
        fun setSpan(sb: SpannableStringBuilder, from: Int, to: Int, spans: Span?) {
            if (spans != null && to - from > 0) {
                when (spans) {
                    is Array<*> if spans.isArrayOf<Span>() -> {
                        for (span in spans) {
                            if (span != null) {
                                sb.setSpan(span, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                            }
                        }
                    }

                    is Collection<*> -> {
                        for (span2 in spans) {
                            sb.setSpan(span2, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                        }
                    }

                    else -> {
                        sb.setSpan(spans, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                    }
                }
            }
        }

        /**
         * Apply spans
         *
         * @param sb        spannable string builder
         * @param from      start
         * @param to        finish
         * @param flags     flags
         * @param factories span factories to call to get spans
         */
        fun setSpan(sb: SpannableStringBuilder, from: Int, to: Int, flags: Long, vararg factories: SpanFactory) {
            if (to - from > 0) {
                for (spanFactory in factories) {
                    val spans = spanFactory.make(flags)
                    setSpan(sb, from, to, spans)
                }
            }
        }

        // I M A G E

        /**
         * Append spans
         *
         * @receiver spannable string builder
         * @param spans image span with possible image style span
         */
        @ReturnThis
        private fun SpannableStringBuilder.appendImageSpans(vararg spans: Span): SpannableStringBuilder {
            val from = length
            append(COLLAPSEDCHAR)
            val to = length
            for (span in spans) {
                setSpan(span, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
            return this
        }

        /**
         * Append image
         *
         * @receiver spannable string builder
         * @param drawable drawable to use
         */
        @ReturnThis
        fun SpannableStringBuilder.appendImage(drawable: Drawable): SpannableStringBuilder {
            val span: Span = ImageSpan(drawable, DynamicDrawableSpan.ALIGN_BASELINE)
            appendImageSpans(span)
            return this
        }

        // C L I C K A B L E

        // how to use:
        // val textView: TextView = ...
        // prepareTextViewForClickableSpan(tv)
        // textView.text= SpannableStringBuilder()
        //  .appendClickableText("text", callback)

        /**
         * Append clickable text
         *
         * @receiver       spannable string builder
         * @param text     text
         * @param listener click listener
         */
        @Suppress("unused")
        @ReturnThis
        fun SpannableStringBuilder.appendClickableText(text: CharSequence, listener: () -> Unit): SpannableStringBuilder {
            val span: ClickableSpan = object : ClickableSpan() {

                override fun onClick(view: View) {
                    Log.d(TAG, "Click text $text")
                    listener()
                }

                override fun updateDrawState(drawState: TextPaint) {
                    super.updateDrawState(drawState)
                    drawState.isUnderlineText = true // Optional: underline the text
                    drawState.color = Color.RED // Optional: change text color
                }
            }
            appendWithSpans(text, span)
            return this
        }

        @Suppress("unused")
        fun prepareTextViewForClickableSpan(tv: TextView) {
            tv.movementMethod = LinkMovementMethod.getInstance()
            tv.isFocusable = true
            tv.isFocusableInTouchMode = true
            tv.isClickable = true
            tv.isLongClickable = true
            tv.isLongClickable = true
            tv.setOnClickListener {}
        }

        /**
         * Append clickable image
         *
         * @receiver       spannable string builder
         * @param caption  caption
         * @param listener click listener
         * @param context  context
         */
        @ReturnThis
        fun SpannableStringBuilder.appendClickableImage(caption: CharSequence, listener: (sb: SpannableStringBuilder?, position: Int, collapsed: Boolean) -> Unit, context: Context): SpannableStringBuilder {
            val collapsedDrawable = getDrawable(context, R.drawable.ic_collapsed)
            val expandedDrawable = getDrawable(context, R.drawable.ic_expanded)
            appendClickableImage(collapsedDrawable, expandedDrawable, caption, listener)
            return this
        }

        /**
         * Append clickable image
         *
         * @receiver                spannable string builder
         * @param collapsedDrawable collapse drawable
         * @param expandedDrawable  expandContainer drawable
         * @param caption           caption
         * @param listener          click listener
         */
        @ReturnThis
        private fun SpannableStringBuilder.appendClickableImage(collapsedDrawable: Drawable, expandedDrawable: Drawable, caption: CharSequence, listener: (sb: SpannableStringBuilder?, position: Int, collapsed: Boolean) -> Unit): SpannableStringBuilder {
            val span = ImageSpan(collapsedDrawable, DynamicDrawableSpan.ALIGN_BASELINE)
            val span2: ClickableSpan = object : ClickableSpan() {
                @Synchronized
                override fun onClick(view: View) {
                    // Log.d(TAG, "Click image")
                    val textView = view as TextView
                    val sb1 = textView.text as SpannableStringBuilder
                    val clickableStart = textView.selectionStart
                    val clickableEnd = textView.selectionEnd
                    val spans: Array<out ImageSpan>? = sb1.getSpans(clickableStart, clickableEnd, ImageSpan::class.java)
                    if (spans != null) {
                        for (span3 in spans) {
                            // get image span
                            val from = sb1.getSpanStart(span3)
                            val to = sb1.getSpanEnd(span3)

                            // remove image span
                            sb1.removeSpan(span3)

                            // text
                            val c = sb1[from]
                            val collapsed = c == COLLAPSEDCHAR
                            sb1.replace(from, to, if (collapsed) EXPANDEDSTRING else COLLAPSEDSTRING)

                            // set new image span
                            val newImageSpan: Span = ImageSpan(if (collapsed) expandedDrawable else collapsedDrawable, DynamicDrawableSpan.ALIGN_BASELINE)
                            sb1.setSpan(newImageSpan, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)

                            // fire click
                            Log.d(TAG, "$from->$to")
                            listener(sb1, to + caption.length + 2, collapsed)
                        }
                    }
                    textView.text = sb1
                }
            }
            appendImageSpans(span, span2)
            append(' ').append(caption).append('\n')
            return this
        }

        /**
         * Insert tag
         *
         * @param sb       spannable string builder
         * @param position insert position
         * @param tag      tag
         */
        fun insertTag(sb: SpannableStringBuilder, position: Int, tag: CharSequence) {
            val insert = "$tag\n$EOEXPANDEDSTRING"
            sb.insert(position, insert)
            val mark = position + tag.length + 1
            sb.setSpan(HiddenSpan(), mark, mark + 1, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        }

        /**
         * Collapse
         *
         * @param sb       spannable string builder
         * @param position position to start from (to end-of-expanded string)
         */
        fun collapse(sb: SpannableStringBuilder, position: Int) {
            sb.delete(position, sb.find(position, EOEXPANDEDSTRING))
        }

        // T E X T

        /**
         * Append text
         *
         * @param sb        spannable string builder
         * @param text      text
         * @param flags     flags
         * @param factories span factories
         * @return input spannable string builder
         */
        fun append(sb: SpannableStringBuilder, text: CharSequence?, flags: Long, vararg factories: SpanFactory?): Appendable {
            if (!text.isNullOrEmpty()) {
                val from = sb.length
                sb.append(text)
                val to = sb.length
                for (spanFactory in factories) {
                    val span = spanFactory!!.make(flags)
                    if (span != null) {
                        sb.applySpan(from, to, span)
                    }
                }
            }
            return sb
        }

        /**
         * Append text
         *
         * @receiver    spannable string builder
         * @param text  text
         * @param spans spans to apply
         */
        @ReturnThis
        fun SpannableStringBuilder.appendWithSpans(text: CharSequence?, vararg spans: Span): SpannableStringBuilder {
            if (!text.isNullOrEmpty()) {
                val from = length
                append(text)
                val to = length
                applySpans(from, to, *spans)
            }
            return this
        }

        /**
         * Apply spans
         *
         * @receiver    spannable string builder
         * @param from  from position
         * @param to    to position
         * @param spans spans to apply
         */
        @ReturnThis
        fun SpannableStringBuilder.applySpans(from: Int, to: Int, vararg spans: Span): SpannableStringBuilder {
            for (span in spans) {
                applySpan(from, to, span)
            }
            return this
        }

        /**
         * Apply span
         *
         * receiver   spannable string builder
         * @param from from position
         * @param to   to position
         * @param span span to apply
         */
        @ReturnThis
        private fun SpannableStringBuilder.applySpan(from: Int, to: Int, span: Span): SpannableStringBuilder {
            when (span) {
                is Array<*> if span.isArrayOf<Span>() -> {
                    for (span2 in span) {
                        setSpan(span2, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                    }
                }

                is Collection<*> -> {
                    for (span2 in span) {
                        setSpan(span2, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                    }
                }

                else -> {
                    setSpan(span, from, to, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                }
            }
            return this
        }

        // H E L P E R S

        /**
         * Find delimiter
         *
         * @receiver        spannable string builder
         * @param start     search start
         * @param delimiter delimiter
         * @return delimiter position or -1 if not found
         */
        private fun SpannableStringBuilder.find(start: Int, @Suppress("SameParameterValue") delimiter: Char): Int {
            var i = start
            while (i < length) {
                if (this[i] == delimiter) {
                    return i + 1
                }
                i++
            }
            return -1
        }

        /**
         * Get drawable from resource id
         *
         * @param activityContext context
         * @param resId resource id
         * @return drawable
         */
        fun getDrawable(activityContext: Context, @DrawableRes resId: Int): Drawable {
            val drawable = AppCompatResources.getDrawable(activityContext, resId)!!
            drawable.setBounds(0, 0, drawable.intrinsicWidth, drawable.intrinsicHeight)
            return drawable
        }
    }
}
